packages/xl-ai/src/prosemirror/changeset.ts

```diff
@@ -0,0 +1,340 @@
+  // expected: hello_world_!
+  //
+  // actual:   hello_world_!
+  // expected: helloworld!
+  //
+  // note: might be possible to merge this loop with the one at the start of the function
+  // but considered ok for now
+  while (diffStart !== null) {
+    const expectedNode = expectedDoc.resolve(diffStart).nodeAfter;
+    const actualNode = tr.doc.resolve(diffStart).nodeAfter;
+
+    if (!expectedNode || !actualNode) {
+      throw new Error("diffNode not found");
+    }
+
+    const isNodeAttrChange =
+      !expectedNode.isLeaf && expectedNode.content.eq(actualNode.content);
+
+    const length = isNodeAttrChange
+      ? 1
+      : Math.min(expectedNode.nodeSize, actualNode.nodeSize);
+
+    const to = diffStart + length;
+    const fromA = invertMap.map(diffStart);
+    const toA = invertMap.map(to);
+
+    // find the position in changes array to insert the new change
+    let insertPos = changes.length;
+    for (let i = 0; i < changes.length; i++) {
+      if (changes[i].fromA >= toA) {
+        insertPos = i;
+        break;
+      }
+    }
+
+    changes.splice(insertPos, 0, {
+      fromA: fromA,
+      toA: toA,
+      fromB: diffStart,
+      toB: to,
+      deleted: [],
+      inserted: [],
+      type: isNodeAttrChange ? "node-type-or-attr-update" : "mark-update",
+    });
+
+    // apply the step so we can find the next diff
+    // Note: even for node type changes, this works - although maybe a ReplaceAroundStep might be cleaner?
+    tr.step(
+      new ReplaceStep(
+        diffStart,
+        to,
+        expectedDoc.slice(diffStart, to),
+        isNodeAttrChange,
+      ),
+    );
+    const newDiffStart = tr.doc.content.findDiffStart(expectedDoc.content);
+
+    if (newDiffStart === diffStart) {
+      // prevent infinite loop, should not happen
+      throw new Error("diffStart not moving");
+    }
+
+    diffStart = newDiffStart;
+  }
+
+  return changes;
+}
+
+const createEncoder = (doc: Node, updatedDoc: Node) => {
+  // this encoder makes sure unchanged table cells stay intact,
+  // without this, prosemirror-changeset would too eagerly
+  // return changes across table cells (this is covered in test cases).
+
+  const tableCellsOld = new Set<string>();
+  const tableCellsNew = new Set<string>();
+  doc.descendants((node) => {
+    if (node.type.name === "tableCell") {
+      tableCellsOld.add(JSON.stringify(node.toJSON()));
+    }
+  });
+
+  updatedDoc.descendants((node) => {
+    if (node.type.name === "tableCell") {
+      tableCellsNew.add(JSON.stringify(node.toJSON()));
+    }
+  });
+
+  const tableCells = new Set(
+    [...tableCellsOld].filter((cell) => tableCellsNew.has(cell)),
+  );
+
+  const encoder: TokenEncoder<any> = {
+    encodeCharacter: (char) => char,
+    encodeNodeStart: (node) => {
+      if (node.type.name === "tableCell") {
+        const str = JSON.stringify(node.toJSON());
+        if (tableCells.has(str)) {
+          return str;
+        }
+        return node.type.name;
```

Comment:
Maybe worth a comment that you are encoding it like this to "flatten" the changeset

Reply:
The idea here was to give two equal table cells in before / after a unique Encoding, to nudge prosemirror-changeset to keep them intact.

I think this encoder improved the table handling a bit, but I'm still running into issues like described here: https://github.com/ProseMirror/prosemirror-changeset/issues/22

will require more research unless you have a good idea. At least, I'll add more documentation to the code here